pkg/storegateway/bucket.go

```diff
@@ -617,92 +620,76 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storegatewaypb.Stor
 	}
 	defer done()
 
+	// Send hints about the blocks loaded for this query before sending series or stats.
+	resHints := buildSeriesResponseHints(blocks)
+	if err = s.sendHints(srv, resHints); err != nil {
+		return err
+	}
+
 	var (
-		streamingIterators *streamingSeriesIterators
-		resHints           = &hintspb.SeriesResponseHints{}
+		streamingSeriesCount int
+		seriesSet            storepb.SeriesSet
+		streamingIterators   *streamingSeriesIterators
+		seriesLoadStart      = time.Now()
+		chunksLimiter        = s.chunksLimiterFactory(s.metrics.queriesDropped.WithLabelValues("chunks"))
+		seriesLimiter        = s.seriesLimiterFactory(s.metrics.queriesDropped.WithLabelValues("series"))
 	)
-	for _, b := range blocks {
-		resHints.AddQueriedBlock(b.meta.ULID)
 
-		if b.meta.Compaction.Level == 1 && b.meta.Thanos.Source == block.ReceiveSource && !b.queried.Load() {
-			level.Debug(s.logger).Log("msg", "queried non-compacted block", "blockId", b.meta.ULID, "ooo", b.meta.Compaction.FromOutOfOrder())
-		}
-
-		b.queried.Store(true)
-	}
-	if err := s.sendHints(srv, resHints); err != nil {
+	seriesSet, streamingIterators, err = s.createIteratorForChunksStreamingLabelsPhase(ctx, req, blocks, indexReaders, shardSelector, matchers, chunksLimiter, seriesLimiter, stats)
+	if err != nil {
 		return err
 	}
 
-	streamingSeriesCount := 0
-	if req.StreamingChunksBatchSize > 0 {
-		var (
-			seriesSet       storepb.SeriesSet
-			seriesLoadStart = time.Now()
-			chunksLimiter   = s.chunksLimiterFactory(s.metrics.queriesDropped.WithLabelValues("chunks"))
-			seriesLimiter   = s.seriesLimiterFactory(s.metrics.queriesDropped.WithLabelValues("series"))
-		)
-
-		seriesSet, streamingIterators, err = s.createIteratorForChunksStreamingLabelsPhase(ctx, req, blocks, indexReaders, shardSelector, matchers, chunksLimiter, seriesLimiter, stats)
-		if err != nil {
-			return err
-		}
-
-		streamingSeriesCount, err = s.sendStreamingSeriesLabelsAndStats(req, srv, stats, seriesSet)
-		if err != nil {
-			return err
-		}
-		spanLogger.DebugLog(
-			"msg", "sent streaming series",
-			"num_series", streamingSeriesCount,
-			"duration", time.Since(seriesLoadStart),
-		)
+	streamingSeriesCount, err = s.sendStreamingSeriesLabelsAndStats(req, srv, stats, seriesSet)
+	if err != nil {
+		return err
+	}
+	spanLogger.DebugLog(
+		"msg", "sent streaming series",
+		"num_series", streamingSeriesCount,
+		"duration", time.Since(seriesLoadStart),
+	)
 
-		if streamingSeriesCount == 0 {
-			// There is no series to send chunks for.
-			return nil
-		}
+	if streamingSeriesCount == 0 {
+		// There are no series to send chunks for.
+		return nil
 	}
 
-	// We create the limiter twice in the case of streaming so that we don't double count the series
-	// and hit the limit prematurely.
-	chunksLimiter := s.chunksLimiterFactory(s.metrics.queriesDropped.WithLabelValues("chunks"))
-	seriesLimiter := s.seriesLimiterFactory(s.metrics.queriesDropped.WithLabelValues("series"))
+	if !req.SkipChunks {
```

Comment:
Given we don't do anything other than return after this `if` block, what if we invert the condition and return early if `req.SkipChunks` is true?

Reply:
Fixed in abfbb24b0109fe62f7f88b8d456fa02347d998bb